scio-elasticsearch/es8/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,791 @@
+import org.apache.beam.sdk.util.Sleeper;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PDone;
+import org.apache.commons.io.output.CountingOutputStream;
+import org.apache.commons.io.output.NullOutputStream;
+import org.apache.http.HttpHost;
+import org.apache.http.auth.AuthScope;
+import org.apache.http.auth.UsernamePasswordCredentials;
+import org.apache.http.client.CredentialsProvider;
+import org.apache.http.impl.client.BasicCredentialsProvider;
+import org.elasticsearch.client.RestClient;
+import org.elasticsearch.client.RestClientBuilder;
+import org.joda.time.Duration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ElasticsearchIO {
+
+  public static class Write {
+
+    private static final Logger LOG = LoggerFactory.getLogger(Write.class);
+    private static final String RETRY_ATTEMPT_LOG =
+        "Error writing to Elasticsearch. Retry attempt[%d]";
+    private static final String RETRY_FAILED_LOG =
+        "Error writing to ES after %d attempt(s). No more attempts allowed";
+
+    /**
+     * Returns a transform for writing to the Elasticsearch cluster for the given nodes.
+     *
+     * @param nodes addresses for the Elasticsearch cluster
+     */
+    public static <T> Bound<T> withNodes(HttpHost[] nodes) {
+      return new Bound<T>().withNodes(nodes);
+    }
+
+    /**
+     * Returns a transform for writing to Elasticsearch cluster by providing slight delay specified
+     * by flushInterval.
+     *
+     * @param flushInterval delay applied to buffer elements. Defaulted to 1 seconds.
+     */
+    public static <T> Bound withFlushInterval(Duration flushInterval) {
+      return new Bound<T>().withFlushInterval(flushInterval);
+    }
+
+    /**
+     * Returns a transform for writing to Elasticsearch cluster.
+     *
+     * @param function creates IndexRequest required by Elasticsearch client
+     */
+    public static <T> Bound withFunction(
+        SerializableFunction<T, Iterable<BulkOperation>> function) {
+      return new Bound<T>().withFunction(function);
+    }
+
+    /**
+     * Returns a transform for writing to Elasticsearch cluster. Note: Recommended to set this
+     * number as number of workers in your pipeline.
+     *
+     * @param numOfShard to construct a batch to bulk write to Elasticsearch.
+     */
+    public static <T> Bound withNumOfShard(long numOfShard) {
+      return new Bound<>().withNumOfShard(numOfShard);
+    }
+
+    /**
+     * Returns a transform for writing to Elasticsearch cluster.
+     *
+     * @param error applies given function if specified in case of Elasticsearch error with bulk
+     *     writes. Default behavior throws IOException.
+     */
+    public static <T> Bound withError(ThrowingConsumer<BulkExecutionException> error) {
+      return new Bound<>().withError(error);
+    }
+
+    public static <T> Bound withMaxBulkRequestSize(int maxBulkRequestSize) {
```

Comment:
add Javadoc for this + `withMaxBulkRequestBytes` and explain how they're used differently?